diff --git a/builtins/ulimit.def b/builtins/ulimit.def
index 62def4f..71eeecc 100644
--- a/builtins/ulimit.def
+++ b/builtins/ulimit.def
@@ -598,7 +598,7 @@ set_limit (ind, newlim, mode)
       }
   else
     {
-#if defined (HAVE_RESOURCE)
+#if defined (HAVE_RESOURCE) && !defined(__redox__)
       if (getrlimit (limits[ind].parameter, &limit) < 0)
 	return -1;
 #  if defined (HPUX9)
diff --git a/config-top.h b/config-top.h
index d89682e..382f2c6 100644
--- a/config-top.h
+++ b/config-top.h
@@ -63,14 +63,14 @@
 /* The default value of the PATH variable. */
 #ifndef DEFAULT_PATH_VALUE
 #define DEFAULT_PATH_VALUE \
-  "/usr/local/bin:/usr/local/sbin:/usr/bin:/usr/sbin:/bin:/sbin:."
+  "/usr/local/bin;/usr/local/sbin;/usr/bin;/usr/sbin;/bin;/sbin;."
 #endif
 
 /* The value for PATH when invoking `command -p'.  This is only used when
    the Posix.2 confstr () function, or CS_PATH define are not present. */
 #ifndef STANDARD_UTILS_PATH
 #define STANDARD_UTILS_PATH \
-  "/bin:/usr/bin:/sbin:/usr/sbin:/etc:/usr/etc"
+  "/bin;/usr/bin;/sbin;/usr/sbin;/etc;/usr/etc"
 #endif
 
 /* Default primary and secondary prompt strings. */
diff --git a/configure.ac b/configure.ac
index ce4e9b6..b39a611 100644
--- a/configure.ac
+++ b/configure.ac
@@ -90,6 +90,7 @@ sparc-linux*)	opt_bash_malloc=no ;;	# sparc running linux; requires ELF
 *-opennt*|*-interix*)	opt_bash_malloc=no ;;	# Interix, now owned by Microsoft
 *-nsk*)		opt_bash_malloc=no ;;	# HP NonStop
 *-haiku*)	opt_bash_malloc=no ;;	# Haiku OS
+*-redox*)	opt_bash_malloc=no ;;	# Redox OS
 esac
 
 # memory scrambling on free()
diff --git a/execute_cmd.c b/execute_cmd.c
index 2a3df6d..0ceb839 100644
--- a/execute_cmd.c
+++ b/execute_cmd.c
@@ -1335,14 +1335,16 @@ time_command (command, asynchronous, pipe_in, pipe_out, fds_to_close)
   nullcmd = (command == 0) || (command->type == cm_simple && command->value.Simple->words == 0 && command->value.Simple->redirects == 0);
   if (posixly_correct && nullcmd)
     {
-#if defined (HAVE_GETRUSAGE)
+#if defined (HAVE_GETRUSAGE) && defined (HAVE_GETTIMEOFDAY)
       selfb.ru_utime.tv_sec = kidsb.ru_utime.tv_sec = selfb.ru_stime.tv_sec = kidsb.ru_stime.tv_sec = 0;
       selfb.ru_utime.tv_usec = kidsb.ru_utime.tv_usec = selfb.ru_stime.tv_usec = kidsb.ru_stime.tv_usec = 0;
       before.tv_sec = shell_start_time;
       before.tv_usec = 0;
 #else
+#  if defined (HAVE_TIMES)
       before.tms_utime = before.tms_stime = before.tms_cutime = before.tms_cstime = 0;
       tbefore = shell_start_time;
+#endif
 #endif
     }
 
diff --git a/general.c b/general.c
index 2d327ad..7d4a6cd 100644
--- a/general.c
+++ b/general.c
@@ -476,6 +476,7 @@ fd_ispipe (fd)
 void
 check_dev_tty ()
 {
+#if !defined(__redox__)
   int tty_fd;
   char *tty;
 
@@ -490,6 +491,7 @@ check_dev_tty ()
     }
   if (tty_fd >= 0)
     close (tty_fd);
+#endif
 }
 
 /* Return 1 if PATH1 and PATH2 are the same file.  This is kind of
@@ -909,10 +911,10 @@ extract_colon_unit (string, p_index)
      `:'.  If I is 0, then the path has a leading colon.  Trailing colons
      are handled OK by the `else' part of the if statement; an empty
      string is returned in that case. */
-  if (i && string[i] == ':')
+  if (i && string[i] == ';')
     i++;
 
-  for (start = i; string[i] && string[i] != ':'; i++)
+  for (start = i; string[i] && string[i] != ';'; i++)
     ;
 
   *p_index = i;
diff --git a/include/posixwait.h b/include/posixwait.h
index 815ea22..8c5e818 100644
--- a/include/posixwait.h
+++ b/include/posixwait.h
@@ -34,7 +34,7 @@
 
 /* How to get the status of a job.  For Posix, this is just an
    int, but for other systems we have to crack the union wait. */
-#if !defined (_POSIX_VERSION)
+#if 0
 typedef union wait WAIT;
 #  define WSTATUS(t)  (t.w_status)
 #else /* _POSIX_VERSION */
@@ -50,7 +50,7 @@ typedef int WAIT;
 
 /* More Posix P1003.1 definitions.  In the POSIX versions, the parameter is
    passed as an `int', in the non-POSIX version, as `union wait'. */
-#if defined (_POSIX_VERSION)
+#if 1
 
 #  if !defined (WSTOPSIG)
 #    define WSTOPSIG(s)       ((s) >> 8)
diff --git a/lib/sh/getcwd.c b/lib/sh/getcwd.c
index d7bd241..779afeb 100644
--- a/lib/sh/getcwd.c
+++ b/lib/sh/getcwd.c
@@ -20,7 +20,7 @@
 
 #include <config.h>
 
-#if !defined (HAVE_GETCWD)
+#if !defined (HAVE_GETCWD) && !defined(__redox__)
 
 #if !defined (__GNUC__) && !defined (HAVE_ALLOCA_H) && defined (_AIX)
   #pragma alloca
diff --git a/sig.c b/sig.c
index ad01631..fe2e253 100644
--- a/sig.c
+++ b/sig.c
@@ -680,7 +680,9 @@ sigterm_sighandler (sig)
 }
 
 /* Signal functions used by the rest of the code. */
-#if !defined (HAVE_POSIX_SIGNALS)
+#if defined(__redox__)
+// Redox already has sigprocmask
+#elif !defined (HAVE_POSIX_SIGNALS)
 
 /* Perform OPERATION on NEWSET, perhaps leaving information in OLDSET. */
 sigprocmask (operation, newset, oldset)
